driver/src/driver/scheduler/system.rs

```diff
@@ -62,42 +65,34 @@ impl<'a> SystemScheduler<'a> {
         }
     }
 
-    /// Returns how long to sleep before starting the next iteration.
-    fn run_once(&mut self, now: SystemTime) -> Duration {
-        const RETRY_TIMEOUT: Duration = Duration::from_secs(10);
-        match self.determine_action(now) {
-            Ok(Action::Sleep(duration)) => {
-                info!("Sleeping {}s.", duration.as_secs());
-                duration
-            }
-            Ok(Action::Solve(batch_id, time_limit)) => {
-                info!(
-                    "Starting to solve batch {} with time limit {}s.",
-                    batch_id.0,
-                    time_limit.as_secs_f64()
-                );
-                match self.driver.run(batch_id.0.into(), time_limit) {
-                    DriverResult::Ok => {
-                        info!("Batch {} solved successfully.", batch_id.0);
-                        self.last_solved_batch.replace(batch_id);
-                        Duration::from_secs(0)
-                    }
-                    DriverResult::Retry(err) => {
-                        error!("StableXDriver retryable error: {}", err);
-                        RETRY_TIMEOUT
-                    }
-                    DriverResult::Skip(err) => {
-                        error!("StableXDriver unretryable error: {}", err);
-                        self.last_solved_batch.replace(batch_id);
-                        Duration::from_secs(0)
+    fn start_solving_in_thread<'b>(
+        &self,
+        batch_id: BatchId,
+        mut solver_time_limit: Duration,
```

Comment:
nit: maybe it would make things slightly less coupled if we didn't pass in the solver time but made it the task of the thread to decide how long it has to solve (if it's None/0 it terminates). This way we would take a small chunk of the logic that is currently shared between threads (and mutated below) and isolate it in one place.

Reply:
Thanks to your comment I realized this could be simplified a lot by passing a deadline `Instant` instead of a timeout `Duration`. Now there is no `mut` and there is a clear condition on which the loop exits.